construction_work_items/data/ir_actions_server.xml

```diff
@@ -0,0 +1,130 @@
+        <field name="code"><![CDATA[
+for wi in records:
+    new_product = env['product.product'].create({
+        'name': wi.x_name,
+        'default_code': wi.x_reference,
+        'type': 'service',
+        'service_policy': "delivered_manual",
+        'list_price': wi.x_unit_price,
+        'standard_price': wi.x_unit_cost,
+        'uom_id': wi.x_unit_id.id,
+    })
+    record.write({'x_product_id': new_product.id})
+]]></field>
+    </record>
+    <record id="server_action_create_or_modify_wi_on_sol" model="ir.actions.server">
+        <field name="name">Construction: Work Item - Create/modify work item on sale order line</field>
+        <field name="model_id" ref="x_work_item_model" />
+        <field name="state">code</field>
+        <field name="code"><![CDATA[
+if record.x_sale_order_line_id:
+    record.x_sale_order_line_id.write({
+        'price_unit': record.x_unit_price,
+        'purchase_price': record.x_unit_cost,
+    })
+elif record.x_targeted_so:
+    env['ir.actions.server'].browse(env.ref('construction_work_items.server_action_create_product_for_wi').id).run()
+    env['sale.order.line'].create({
+        'order_id': record.x_targeted_so.id,
+        'product_id': record.x_product_id.id,
+        'price_unit': record.x_unit_price,
+        'purchase_price': record.x_unit_cost,
+        'x_line_work_item_ids': [record.id],
+    })
+]]></field>
+    </record>
+    <record id="server_action_create_associated_work_item" model="ir.actions.server">
+        <field name="name">Construction: SOL - Create associated Work Item</field>
+        <field name="model_id" ref="sale.model_sale_order_line" />
+        <field name="state">code</field>
+        <field name="code"><![CDATA[
+if not record.x_line_work_item_ids and record.product_id and record.product_id.x_work_item_template_id:
+    record.product_id.x_work_item_template_id.copy({
+        'x_sale_order_line_id': record.id,
+        'x_is_template': False,
+    })
+]]></field>
+    </record>
+    <record id="server_action_open_sol_work_item" model="ir.actions.server">
+        <field name="name">Construction: SOL - Open SOL Work Item</field>
+        <field name="model_id" ref="sale.model_sale_order_line" />
+        <field name="state">code</field>
+        <field name="code"><![CDATA[
+if record.x_line_work_item_ids:
+    action = {
+        "type": "ir.actions.act_window",
+        "res_model": "x_work_item",
+        "res_id": record.x_line_work_item_ids.id,
+        "view_mode": "form",
+        "target": "new",
+    }
+]]></field>
+    </record>
+    <record id="server_action_assign_x_template_work_item_id" model="ir.actions.server">
+        <field name="name">Construction: Product - Assign Template Work Item</field>
+        <field name="model_id" ref="x_work_item_model" />
+        <field name="state">code</field>
+        <field name="code"><![CDATA[
+for work_item in records.filtered('x_is_template').filtered('x_product_id').filtered('x_active'):
```

Comment:
each filtered means a loop over each record, better group them, or early continue if one of these conditions is not matched